src/input.rs

```diff
@@ -203,6 +225,106 @@ impl Input {
             MouseScrollDelta::PixelDelta(_, y) => y,
         });
     }
+
+    pub fn button_hits(
+        &self,
+        button: Button,
+    ) -> u8 {
+        use std::u8::MAX;
+        match button {
+            Button::Key(button) => self.1
+                .keys_hit
+                .iter()
+                .filter(|&&key| key == button)
+                .take(MAX as usize)
+                .count() as u8,
+            Button::Mouse(button) => self.1
+                .mouse_hit
+                .iter()
+                .filter(|&&key| key == button)
+                .take(MAX as usize)
+                .count() as u8,
+        }
+    }
+
+    pub fn button_down(
+        &self,
+        button: Button,
+    ) -> bool {
+        match button {
+            Button::Key(button) => self.0.keys_pressed.contains(&button),
+            Button::Mouse(button) => self.0.mouse_pressed.contains(&button),
+        }
+    }
+
+    pub fn axis_hits(
+        &self,
+        axis: Axis,
+    ) -> Option<i8> {
+        match axis {
+            Axis::Key { pos, neg } => {
+                let (mut pos_count, mut neg_count) = (0i8, 0i8);
+                for &key in self.1.keys_hit.iter() {
+                    if key == neg {
+                        neg_count += 1;
+                    }
+                    if key == pos {
+                        pos_count += 1;
+                    }
+                }
+                if pos_count + neg_count == 0 {
+                    None
+                } else {
+                    Some(pos_count + neg_count)
+                }
+            }
+            Axis::Custom(axis_id) => {
+                let mut count = 0i8;
+                for &axis in self.1.axes_raw.iter() {
+                    if axis.0 == axis_id {
+                        count += 1;
+                    }
+                }
+                if count == 0 {
+                    None
+                } else {
+                    Some(count)
+                }
+            }
+        }
+    }
+
+    pub fn axis_delta(
+        &self,
+        axis: Axis,
+    ) -> Option<f32> {
+        match axis {
+            Axis::Key { pos, neg } => {
+                let is_pos = self.0.keys_pressed.contains(&pos);
+                let is_neg = self.0.keys_pressed.contains(&neg);
+                if is_pos && !is_neg {
+                    Some(1.0f32)
+                } else if is_neg && !is_pos {
+                    Some(-1.0f32)
+                } else {
+                    None
+                }
+            }
+            Axis::Custom(axis_id) => {
+                let moves = self.1
+                    .axes_raw
+                    .iter()
+                    .filter(|&&(id, _)| id == axis_id)
+                    .map(|&(_, value)| value)
+                    .collect::<Vec<_>>();
```

Comment:
strictly speaking, this doesn't need to be collected into a `Vec` since all you need is to iterate it here for the `sum`

Reply:
I know, but I need to return `None` in case of no elements in this vector. I'm not sure is it possible without `collect`

Reply:
You could do something like
```rust
let iterator = ...;
if iterator.clone().count() != 0 {
   iterator.sum()
} else {
  None
}
```